### PR TITLE
AOTAutograd: keep set_() input mutations in the graph, ban other cases

### PR DESCRIPTION
**Pull request by bdhirsh**
We have some (limited) support for `set_()` input mutations in `torch.compile`, but one restriction today is that we force them to run outside of the graph, in the opaque runtime epilogue.

This is a problem for ppFSDP. Why? The usage pattern of ppFSDP forward graphs look something like this:
```
def forward_fsdp(sacrificial_param, sharded_param, inp):
    allgathered_param = allgather(sharded_param)
    sacrificial_param.set_(allgathered_param)  # hidden in an autograd.Function that we trace
    out = matmul(sacrificial_param, inp)
    sacrificial_param.untyped_storage().resize_(0)
    return out
```
When we functionalize this graph, `sacrificial_param` sees two distinct types of input mutations, that we must preserve: a `set_`, and a `resize_`. Importantly, at runtime the `set_()` must run **before** the `resize_()`. Why? the `set_()` updates the storage of our sacrificial param to the allgather'd data, which allows the call to `sacrificial_param.resize_()` to free the allgathered data later. If we run the two mutations in reverse order, we will never free the allgathered data.

We want to put the `resize_()` mutation op inside of the graph (see next PR, also there's a much longer description in that PR for anyone interested). However, this will require us to put `set_()` in the graph as well, in order for them to run in the correct order.

In order to do this, I had to add some extra restrictions: You are now required to run `set_()` under `no_grad()` if you use it with `torch.compile`, and if you perform any other mutations to the input, those must be under no_grad as well (otherwise, the mutations may mutate the `grad_fn` of the input, making it no longer safe to keep in the graph). These restrictions are hopefully reasonable, since `set_()` doesn't see much usage today (and the original impetus for adding set_() support a few months ago was for fsdp anyway)

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #122434
* __->__ #122981
* #123646
* #122433

### CODE REVIEW

**Review by jansel — APPROVED** [review]

**Comment by bdhirsh** [comment]
Still working through CI failures on this stack. I fixed 3 `set_()` issues:

- Putting `set_()` in the graph broke our handling for detecting if we are returning graph inputs as outputs
- We were previously `.detach()`ing inputs before calling the compiled fw **whenever** we detected that they had None gradients. We don't have to do this if the input already doesn't require grad, and we cannot do this if the input had a `set_()`. This should also improve some of the fixed overheads in AOTAutograd
- Dumb bug we had previously never ran into when there was a storage mutation in the graph and we needed to return an input directly: https://github.com/pytorch/pytorch/pull/122981/files#diff-594ef3e84a8e3555bfa4a6167383949da5f321e93f602246eb9c0a797d5ecc00R536

**Comment by bdhirsh** [comment]
Some of my `detach()` changes were failing torchbench, should be fixed. I think CI should be mostly green now on this PR, will watch the CI for the ppFSDP changes higher up

File: torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py [hunk 1]
```diff
@@ -371,7 +371,18 @@ def aot_dispatch_autograd(
...
+                # If our input experiences a metadata mutation inside the graph (e.g. set_()),
+                # we *must* not detach, otherwise it will be the detach'd input that gets the metadata mutation
+                metadata_mutation_in_graph = (
+                    fw_metadata.input_info[i].mutation_type
+                    == MutationType.MUTATED_IN_GRAPH
+                    and fw_metadata.input_info[i].mutates_storage_metadata
+                )
+                if (
+                    bw_out is None
+                    and not metadata_mutation_in_graph
```
Comment (bdhirsh) [inline torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py:383]: oh fwiw @jansel - this is probably what makes it ok for me to remove the `can_detach()` logic.

I guess it is morally similar to the `can_detach()` code, but it applies for any input metadata mutations that we keep in the graph: if there is a graph input that we mutate the metadata of (via `set_()`, or any other metadata mutation), we can't detach first, or else the user's tensor won't get the metadata mutation.
